lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java

```diff
@@ -106,6 +114,63 @@ final int versionStamp() {
     return map.versionStamp;
   }
 
+  private int registerReference(Persistance.Reference<?> ref) {
+    var obj = ref.get(Object.class);
+    var existingId = pendingReferences.get(obj);
+    if (existingId == null) {
+      var currentSize = countReferences++;
+      pendingReferences.put(obj, currentSize);
+      return currentSize;
+    } else {
+      return existingId;
+    }
+  }
+
+  /**
+   * Writes an object into the buffer. Writes also all {@link Persistance.Reference} that were left
+   * pending during the serialization.
+   *
+   * @param obj the object to write down
+   * @return location of the table {@code int size and then int[size]}
+   */
+  private int writeObjectAndReferences(Object obj) throws IOException {
+    var objAt = writeObject(obj);
+
+    var refsOut = new ByteArrayOutputStream();
+    var refsData = new DataOutputStream(refsOut);
+    refsData.writeInt(-1); // space for size of references
+    refsData.writeInt(objAt); // the main object
```

Comment:
I appreciate such comments a lot. With such a low-level binary format it is really not immediately clear what is what. Either creating small functions like `writePlaceholder() { refsData.writeInt(-1); }` OR adding comments like this helps immensely with understanding this code.

Reply:
> small functions

This PR introduces [putIntToArray](https://github.com/enso-org/enso/pull/9972/files#diff-9d27b6bd14cc82cf4c3817f34013a7b88be171835fc553b522bf61d255bf0680R56), at least.

Reply:
I'm very happy that it was introduced.

Reply:
In my ideal world, we'd try to [split the levels of abstraction](https://www.fluentcpp.com/2016/12/15/respect-levels-of-abstraction/) even more and have separate protocol-level semantic `putMainObjectReference` functions delegating to lower-level byte-buffer manipulation `putIntToArray` and `writeInt` functions.